File: keycloak/user-storage-s2/src/main/java/com/example/keycloak/s2/S2UserAdapter.java
package com.example.keycloak.s2;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

import org.keycloak.component.ComponentModel;
import org.keycloak.credential.CredentialInput;
import org.keycloak.credential.CredentialModel;
import org.keycloak.models.KeycloakSession;
import org.keycloak.models.RealmModel;
import org.keycloak.models.SubjectCredentialManager;
import org.keycloak.models.UserModel;
import org.keycloak.storage.StorageId;
import org.keycloak.storage.adapter.AbstractUserAdapterFederatedStorage;

/**
 * Minimal user adapter that exposes Service2 users to Keycloak.
 */
class S2UserAdapter extends AbstractUserAdapterFederatedStorage.Streams {

    private final String id;
    private final String username;
    private final S2UserStorageProvider provider;
    private final SubjectCredentialManager credentialManager;

    S2UserAdapter(KeycloakSession session, RealmModel realm, ComponentModel model, S2UserStorageProvider provider, String username) {
        super(session, realm, model);
        this.username = username;
        this.provider = provider;
        this.id = StorageId.keycloakId(model, username);
        this.credentialManager = new S2SubjectCredentialManager();
        setFederationLink(model.getId());
    }

    @Override
    public SubjectCredentialManager credentialManager() {
        return credentialManager;
    }

    @Override
    public String getId() {
        return id;
    }

    @Override
    public String getUsername() {
        return username;
    }

    @Override
    public void setUsername(String username) {
        throw new UnsupportedOperationException("Read only user");
    }

    @Override
    public String getEmail() {
        return username + "@service2.local";
    }

    @Override
    public void setEmail(String email) {
        throw new UnsupportedOperationException("Read only user");
    }

    @Override
    public String getFirstName() {
        return "Service2";
    }

    @Override
    public void setFirstName(String firstName) {
        throw new UnsupportedOperationException("Read only user");
    }

    @Override
    public String getLastName() {
        return "User";
    }

    @Override
    public void setLastName(String lastName) {
        throw new UnsupportedOperationException("Read only user");
    }

    @Override
    public boolean isEnabled() {
        return true;
    }

    @Override
    public Map<String, List<String>> getAttributes() {
        Map<String, List<String>> attributes = new HashMap<>();
        attributes.put(UserModel.USERNAME, List.of(getUsername()));
        attributes.put(UserModel.EMAIL, List.of(getEmail()));
        attributes.put(UserModel.FIRST_NAME, List.of(getFirstName()));
        attributes.put(UserModel.LAST_NAME, List.of(getLastName()));
        return Collections.unmodifiableMap(attributes);
    }

    @Override
    public Stream<String> getAttributeStream(String name) {
        return getAttributes().getOrDefault(name, List.of()).stream();
    }

    @Override
    public void setAttribute(String name, List<String> values) {
        throw new UnsupportedOperationException("Read only user");
    }

    @Override
    public void setSingleAttribute(String name, String value) {
        throw new UnsupportedOperationException("Read only user");
    }

    @Override
    public void removeAttribute(String name) {
        throw new UnsupportedOperationException("Read only user");
    }

    private class S2SubjectCredentialManager implements SubjectCredentialManager {

        @Override
        public boolean isValid(List<CredentialInput> inputs) {
            if (inputs == null || inputs.isEmpty()) {
                return false;
            }
            return inputs.stream().allMatch(input -> provider.isValid(realm, S2UserAdapter.this, input));
        }

        @Override
        public boolean updateCredential(CredentialInput input) {
            throw new UnsupportedOperationException("Read only user");
        }

        @Override
        public void updateStoredCredential(CredentialModel cred) {
            throw new UnsupportedOperationException("Read only user");
        }

        @Override
        public CredentialModel createStoredCredential(CredentialModel cred) {
            throw new UnsupportedOperationException("Read only user");
        }

        @Override
        public boolean removeStoredCredentialById(String id) {
            throw new UnsupportedOperationException("Read only user");
        }

        @Override
        public CredentialModel getStoredCredentialById(String id) {
            return null;
        }

        @Override
        public Stream<CredentialModel> getStoredCredentialsStream() {
            return Stream.empty();
        }

        @Override
        public Stream<CredentialModel> getStoredCredentialsByTypeStream(String type) {
            return Stream.empty();
        }

        @Override
        public CredentialModel getStoredCredentialByNameAndType(String name, String type) {
            return null;
        }

        @Override
        public boolean moveStoredCredentialTo(String id, String newPreviousCredentialId) {
            throw new UnsupportedOperationException("Read only user");
        }

        @Override
        public void updateCredentialLabel(String credentialId, String credentialLabel) {
            throw new UnsupportedOperationException("Read only user");
        }

        @Override
        public void disableCredentialType(String credentialType) {
            throw new UnsupportedOperationException("Read only user");
        }

        @Override
        public Stream<String> getDisableableCredentialTypesStream() {
            return Stream.empty();
        }

        @Override
        public boolean isConfiguredFor(String type) {
            return provider.isConfiguredFor(realm, S2UserAdapter.this, type);
        }

        @Override
        public boolean isConfiguredLocally(String type) {
            return isConfiguredFor(type);
        }

        @Override
        public Stream<String> getConfiguredUserStorageCredentialTypesStream() {
            return Stream.empty();
        }

        @Override
        public CredentialModel createCredentialThroughProvider(CredentialModel model) {
            throw new UnsupportedOperationException("Read only user");
        }
    }
}

File: keycloak/user-storage-s2/src/main/kotlin/com/example/keycloak/s2/S2UserAdapter.kt
package com.example.keycloak.s2

import java.util.Collections
import java.util.stream.Stream
import java.util.stream.StreamSupport
import org.keycloak.component.ComponentModel
import org.keycloak.credential.CredentialInput
import org.keycloak.credential.CredentialModel
import org.keycloak.models.KeycloakSession
import org.keycloak.models.RealmModel
import org.keycloak.models.SubjectCredentialManager
import org.keycloak.models.UserModel
import org.keycloak.storage.StorageId
import org.keycloak.storage.adapter.AbstractUserAdapter

/**
 * Minimal user adapter that exposes Service2 users to Keycloak.
 */
class S2UserAdapter(
    session: KeycloakSession,
    realm: RealmModel,
    model: ComponentModel,
    private val provider: S2UserStorageProvider,
    private val username: String
) : AbstractUserAdapter.Streams(session, realm, model) {

    private val id: String = StorageId.keycloakId(model, username)
    private val credentialManager: SubjectCredentialManager = S2SubjectCredentialManager()
    private val attributes: MutableMap<String, List<String>> = mutableMapOf(
        UserModel.USERNAME to listOf(username),
        UserModel.EMAIL to listOf("$username@service2.local"),
        UserModel.FIRST_NAME to listOf("Service2"),
        UserModel.LAST_NAME to listOf("User")
    )

    override fun credentialManager(): SubjectCredentialManager = credentialManager

    override fun getId(): String = id

    override fun getUsername(): String = username

    override fun setUsername(username: String) {
        throw UnsupportedOperationException("Read only user")
    }

    override fun getEmail(): String? =
        attributes[UserModel.EMAIL]?.firstOrNull()

    override fun setEmail(email: String) {
        throw UnsupportedOperationException("Read only user")
    }

    override fun getFirstName(): String? =
        attributes[UserModel.FIRST_NAME]?.firstOrNull()

    override fun setFirstName(firstName: String) {
        throw UnsupportedOperationException("Read only user")
    }

    override fun getLastName(): String? =
        attributes[UserModel.LAST_NAME]?.firstOrNull()

    override fun setLastName(lastName: String) {
        throw UnsupportedOperationException("Read only user")
    }

    override fun getAttributes(): Map<String, List<String>> =
        Collections.unmodifiableMap(attributes)

    override fun getAttributeStream(name: String): Stream<String> =
        attributes[name]?.let { StreamSupport.stream(it.spliterator(), false) } ?: Stream.empty()

    override fun setAttribute(name: String, values: MutableList<String>) {
        throw UnsupportedOperationException("Read only user")
    }

    override fun setSingleAttribute(name: String, value: String) {
        throw UnsupportedOperationException("Read only user")
    }

    override fun removeAttribute(name: String) {
        throw UnsupportedOperationException("Read only user")
    }

    private inner class S2SubjectCredentialManager : SubjectCredentialManager {
        override fun isValid(inputs: MutableList<CredentialInput>?): Boolean {
            if (inputs.isNullOrEmpty()) {
                return false
            }
            return inputs.all { provider.isValid(realm, this@S2UserAdapter, it) }
        }

        override fun updateCredential(input: CredentialInput?): Boolean {
            throw UnsupportedOperationException("Read only user")
        }

        override fun updateStoredCredential(cred: CredentialModel?) {
            throw UnsupportedOperationException("Read only user")
        }

        override fun createStoredCredential(cred: CredentialModel?): CredentialModel {
            throw UnsupportedOperationException("Read only user")
        }

        override fun removeStoredCredentialById(id: String?): Boolean {
            throw UnsupportedOperationException("Read only user")
        }

        override fun getStoredCredentialById(id: String?): CredentialModel? = null

        override fun getStoredCredentialsStream(): Stream<CredentialModel> = Stream.empty()

        override fun getStoredCredentialsByTypeStream(type: String?): Stream<CredentialModel> = Stream.empty()

        override fun getStoredCredentialByNameAndType(name: String?, type: String?): CredentialModel? = null

        override fun moveStoredCredentialTo(id: String?, newPreviousCredentialId: String?): Boolean {
            throw UnsupportedOperationException("Read only user")
        }

        override fun updateCredentialLabel(credentialId: String?, credentialLabel: String?) {
            throw UnsupportedOperationException("Read only user")
        }

        override fun disableCredentialType(credentialType: String?) {
            throw UnsupportedOperationException("Read only user")
        }

        override fun getDisableableCredentialTypesStream(): Stream<String> = Stream.empty()

        override fun isConfiguredFor(type: String?): Boolean =
            provider.isConfiguredFor(realm, this@S2UserAdapter, type)

        override fun isConfiguredLocally(type: String?): Boolean = isConfiguredFor(type)

        override fun getConfiguredUserStorageCredentialTypesStream(): Stream<String> = Stream.empty()

        override fun createCredentialThroughProvider(model: CredentialModel?): CredentialModel {
            throw UnsupportedOperationException("Read only user")
        }
    }
}
